src/java.base/share/classes/java/io/Reader.java

```diff
@@ -184,12 +184,25 @@ protected Reader(Object lock) {
      * @since 1.5
      */
     public int read(CharBuffer target) throws IOException {
-        int len = target.remaining();
-        char[] cbuf = new char[len];
-        int n = read(cbuf, 0, len);
-        if (n > 0)
-            target.put(cbuf, 0, n);
-        return n;
+        int nread;
+        if (target.hasArray()) {
+            char[] cbuf = target.array();
+            int pos = target.position();
+            int rem = target.limit() - pos;
+            if (rem <= 0)
+                return -1;
+            int off = target.arrayOffset() + pos;
+            nread = this.read(cbuf, off, rem);
+            if (nread > 0)
+                target.position(pos + nread);
+        } else {
+            int len = target.remaining();
+            char[] cbuf = new char[len];
+            nread = read(cbuf, 0, len);
+            if (nread > 0)
+                target.put(cbuf, 0, nread);
+        }
+        return nread;
```

Comment:
Thanks for bringing this back to just the heap buffer case. This part looks good now.

Reply:
Agreed.